### PR TITLE
Fix author archives with no posts returning a 404 instead of author.php

### LINKED ISSUE #54: Empty author archives return a 404 instead of author.php
WordPress core uses `get_query_var( 'author' )` in `handle_404()` of `WP` to detect [if a request is a valid author archive request](https://github.com/WordPress/WordPress/blob/2f263fce9961031d6dfd9546bec596893bd592b3/wp-includes/class-wp.php#L694-L698).

WordPress core fetches the `author` query var value [during `get_posts()` in `WP_Query`](https://github.com/WordPress/WordPress/blob/4b0e64aa8c8723781a2428160d3d9c18a0e0d3cd/wp-includes/class-wp-query.php#L2137-L2139) which doesn't happen in `ES_WP_Query_Wrapper` or even later during `reboot_query_vars()` in `ES_WP_Query_Shoehorn`.

This leads to `get_query_var( 'author' )` returning `null` in `handle_404()` and author archives being returned as 404 if no posts are found.

Ref: VIP ZD66699

### PR DESCRIPTION
**Pull request by kasparsd**
Fixes #54.

### CODE REVIEW

File: class-es-wp-query-shoehorn.php [hunk 1]
```diff
@@ -55,11 +55,14 @@ function es_wp_query_shoehorn( &$query ) {
...
 
-		$es_query_vars = $query->query_vars;
 		$query_args = $query->query;
+
+		// Run this query through ES.
+		$es_query_vars = $query->query_vars;
 		$es_query_vars['fields'] = 'ids';
 		$es_query = new ES_WP_Query( $es_query_vars );
 
+		// Make the post query use the post IDs from the ES results instead.
```
Comment (kasparsd) [inline class-es-wp-query-shoehorn.php]: Added some comments to explain why we're adjusting the query.

File: class-es-wp-query-shoehorn.php [hunk 2]
```diff
@@ -218,5 +224,15 @@ private function reboot_query_vars( &$query ) {
 		if ( empty( $q['posts_per_page'] ) ) {
 			$q['posts_per_page'] = get_option( 'posts_per_page' );
 		}
+
+		// Restore the author ID which is normally added during get_posts() in WP_Query.
+		// Required for handle_404() in WP class to not mark empty author archives as 404s.
+		if ( $query->is_author() && ! empty( $q['author_name'] ) ) {
```
Comment (kasparsd) [inline class-es-wp-query-shoehorn.php]: This restores the core functionality from [`get_posts()` in `WP_Query`](https://github.com/WordPress/WordPress/blob/4b0e64aa8c8723781a2428160d3d9c18a0e0d3cd/wp-includes/class-wp-query.php#L2137-L2139).

Comment (ethitter) [reply class-es-wp-query-shoehorn.php]: For consistency with the `get_posts()` method, we should include the handling for author names with slashes in them, which appears just above the lines you've referenced.

**Comment by mboynes** [comment]
I'll address the failing travis build on PHP 7.2 elsewhere

**Pull request merged**